src/cluster_asm.c

```diff
@@ -2449,7 +2453,13 @@ void asmSendStreamEofIfDrained(asmTask *task) {
 }
 
 void asmBeforeSleep(void) {
-    asmTrimJobProcessPending();
+    /* Handle pending trim jobs. */
+    if (clusterNodeIsMaster(getMyClusterNode())) {
+        asmTrimJobProcessPending();
+    } else {
+        /* Cancel pending trim jobs if becoming a replica. */
+        asmCancelPendingTrimJobs();
+    }
```

Comment:
Not sure I understand this part. Why are we cancelling on replica? Just a reminder, master and replica handle trim separately in the current design.

Reply:
We should deal with unowned keys as quick as possible, otherwise it causes duplication for modules as they replicate their queries and merge all the results. So, I also feel like reverting `asmTrimSlotsIfNotOwned()` to trim all the unowned slots, just in case.

Reply:
You are right, 4-a seems not a problem as it will require full sync.

> this change may break legacy slot migration and cause data loss

Remember, now we call `asmTrimSlotsIfNotOwned()` only from ASM functions. Once it is called, we know we are not using legacy migration. So, I think it can trim all the slots rather than the specific ones, no?

Reply:
Btw, if double trim bothers you, maybe we can avoid running trim if the slots are empty. We already calculate keys number of keys to be trimmed anyway.

Reply:
i don't have an idea.  so you accept double trim, right?

Reply:
not sure, just feel uncomfortable

Reply:
revert